cpp/pipelines/augment_canones_references.cc

```diff
@@ -0,0 +1,210 @@
+/** \file    augment_canones_references.cc
+ *  \brief   A tool for adding numerical canon law references to MARC-21 datasets.
+ *  \author  Dr. Johannes Ruscheinski
+ */
+
+/*
+    Copyright (C) 2019, Library of the University of Tübingen
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <cstdlib>
+#include "BibleUtil.h"
+#include "Compiler.h"
+#include "MARC.h"
+#include "StringUtil.h"
+#include "UBTools.h"
+#include "util.h"
+
+
+namespace {
+
+
+void ParseRanges(const std::string &ppn, const std::string &ranges, unsigned * const range_start, unsigned * const range_end) {
+    std::vector<std::string> parts;
+    if (unlikely(StringUtil::SplitThenTrimWhite(ranges, ',', &parts, /* suppress_empty_components = */false) > 3))
+        LOG_ERROR("bad codex parts \"" + ranges + "\"! (PPN: " + ppn + ")");
+
+    unsigned canones;
+    if (parts.size() == 1) {
+        const auto dash_pos(parts[0].find('-'));
+        if (dash_pos == std::string::npos) {
+            if (unlikely(not StringUtil::ToUnsigned(parts[0], &canones)))
+                LOG_ERROR("bad codex canones \"" + parts[0] + "\"! (PPN: " + ppn + ")");
+            *range_start = canones * 10000;
+            *range_end   = canones * 10000 + 9999;
+        } else {
+            if (unlikely(not StringUtil::ToUnsigned(parts[0].substr(0, dash_pos), range_start)
+                         or not StringUtil::ToUnsigned(parts[0].substr(dash_pos + 1), range_end)))
+                LOG_ERROR("bad canones range \"" + parts[0] + "\"! (PPN: " + ppn + ")");
+            *range_start *= 10000;
+            *range_end   *= 10000;
+        }
+
+        return;
+    } else if (parts.size() == 2) {
+        if (unlikely(not StringUtil::ToUnsigned(parts[0], &canones)))
+            LOG_ERROR("bad codex canones \"" + parts[0] + "\"! (PPN: " + ppn + ")");
+
+        const auto dash_pos(parts[1].find('-'));
```

Comment:
Is it possible to have ranges for both Pars? As in, _CIC83 1044,1-2,2-3_ or _CIC83 1044,3,2-5_. Those two cases aren't currently evaluated by the code.